File: doc/Programs/LecturePrograms/programs/NumericalIntegration/cpp/program1.cpp
#include <iostream>
#include "lib.h"
using namespace std;
//     Here we define various functions called by the main program

double int_function(double x);
//   Main function begins here
int main()
{
     int n;
     double a, b;
     cout << "Read in the number of integration points" << endl;
     cin >> n;
     cout << "Read in integration limits" << endl;
     cin >> a >> b;
//   reserve space in memory for vectors containing the mesh points
//   weights and function values for the use of the gauss-legendre
//   method
     double *x = new double [n];
     double *w = new double [n];
//   set up the mesh points and weights
     gauleg(a, b,x,w, n);
//   evaluate the integral with the Gauss-Legendre method
//   Note that we initialize the sum
     double int_gauss = 0.;
     for ( int i = 0;  i < n; i++){
        int_gauss+=w[i]*int_function(x[i]);
     }
//    final output
      cout << "Trapez-rule = " << trapezoidal_rule(a, b,n, &int_function)
           << endl;
      cout << "Simpson's rule = " << simpson(a, b,n, &int_function) 
           << endl;
      cout << "Gaussian quad = " << int_gauss << endl;
      delete [] x;
      delete [] w;
      return 0;
}  // end of main program
//  this function defines the function to integrate
double int_function(double x)
{
  double value = 4./(1.+x*x);
  return value;
} // end of function to evaluate

File: doc/Programs/LecturePrograms/programs/NumericalIntegration/Fortran/program1.f90
!  Simple test program for numerical integration which
!  calls three methods in the program library (you must link
!  the library as well).


PROGRAM int_test
  USE constants
  USE F90library
  IMPLICIT NONE
  INTEGER :: i, n
  REAL(DP) :: a, b, int_trapez, int_gauss, int_simpson
  REAL(DP), ALLOCATABLE, DIMENSION(:) :: x, w

  INTERFACE
     DOUBLE PRECISION FUNCTION  func(x)
       USE constants
       IMPLICIT NONE
       REAL(DP), INTENT(IN) :: x

     END FUNCTION func
  END INTERFACE

  WRITE(*,*) ' Read in number of mesh points'
  READ(*,*) n
  WRITE(*,*) ' Read in integration limits [a,b]'
  READ(*,*) a, b

  !    reserve space in memory for vectors containing the mesh points
  !    weights and function values for the use of the Gauss-Legendre
  !    method

  ALLOCATE ( x(n), w(n) )

  !    Set up the mesh points and weights for Gauss Legendre

  CALL gauleg(a,b,x,w,n)

  !    Integrate using the trapezoidal rule and simpson's method
  !    Note the transfer of a function name to the methods
  CALL trapezoidal_rule(a,b,int_trapez,n,func)
  CALL simpson(a,b,int_simpson,n,func)
  !    Gaussian quadrature

  int_gauss=0.
  DO i=1,n
     int_gauss=int_gauss+w(i)*func(x(i))
  ENDDO

  !    final output
  WRITE (*, *) n, int_trapez, int_simpson, int_gauss
  DEALLOCATE ( x, w)

END PROGRAM int_test

!    The explicit function to be evaluated

DOUBLE PRECISION FUNCTION  func(x)
  USE constants
  IMPLICIT NONE
  REAL(DP), INTENT(IN) :: x
  func=exp(-x)/x

END FUNCTION func
